torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp

```diff
@@ -511,6 +556,113 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
+  }
+}
+
+void ProcessWeightObserverSubgraph::cloneNodeInGraph(
+    Node* node,
+    std::shared_ptr<Graph>& g) {
+  auto* block = g->block();
+  auto value_fn = [&](Value* v) { return updateInputValueInGraph(v, g); };
+
+  auto new_node = block->appendNode(g->createClone(node, value_fn));
+  for (size_t i = 0; i < node->outputs().size(); ++i) {
+    auto oo = node->outputs()[i];
+    auto no = new_node->outputs()[i];
+    remap_values_in_observer_subgraph_[oo] = no;
+  }
+}
+void ProcessWeightObserverSubgraph::buildObserverSubgraph(
+    const std::vector<Node*>& weight_subgraph,
+    std::shared_ptr<Graph> dest_graph) {
+  // Build weight subgraph
+  for (auto n : weight_subgraph) {
+    cloneNodeInGraph(n, dest_graph);
+  }
+  LintGraph(dest_graph);
+
+  // Add last node output value as subgraph output.
+  for (auto out : weight_subgraph.back()->outputs()) {
+    dest_graph->registerOutput(updateInputValueInGraph(out, dest_graph));
+  }
+  GRAPH_DUMP("New weight observer subgraph: ", dest_graph);
+}
+
+void ProcessWeightObserverSubgraph::findSubgraph(
+    Value* self,
+    Value* input_val,
+    std::vector<Node*>& weight_subgraph) {
+  const auto& inputs = input_val->node()->inputs().vec();
+  for (auto v : inputs) {
+    auto n = v->node();
+    if (!hitGraphInput(n->output(0))) {
+      weight_subgraph.push_back(n);
+      findSubgraph(self, v, weight_subgraph);
+    } else {
+      TORCH_CHECK(
+          n->output(0) == self,
+          "Unexpected value found when handling weight value "
+          " in findSubgraph, traced back to:",
+          n->output(0)->debugName(),
+          " which is not self:",
+          self->debugName());
+    }
+  }
+}
+
+std::unique_ptr<GraphFunction> ProcessWeightObserverSubgraph::
+    buildGraphFromNodes(
+        const std::vector<Node*>& nodes,
+        const std::string& name) {
+  auto observer_subgraph = std::make_shared<Graph>();
+  auto build_observer_graph = [&](Function& func) {
+    buildObserverSubgraph(nodes, func.graph());
+  };
+  return torch::make_unique<GraphFunction>(
+      name, observer_subgraph, build_observer_graph);
+}
+
+void InsertQuantDeQuantHelper::extractAndRunWeightObserver(
+    Module& module,
+    Value* self,
+    Value* weight_value,
+    std::unordered_map<Value*, std::vector<Node*>>& weight_to_subgraph_nodes) {
+  Node* observer = weight_value->node();
+  Graph* g = observer->owningGraph();
+  std::vector<Node*> weight_subgraph;
+  ProcessWeightObserverSubgraph o;
+
+  // If the graph was already visited, return list of relevant nodes directly.
+  // Multiple module instances can share the same graph code, so we don't need
+  // to re-run the extraction process.
+  if (weight_to_subgraph_nodes.count(weight_value)) {
+    weight_subgraph = weight_to_subgraph_nodes[weight_value];
+  } else {
+    // Extract the subgraph nodes.
+    weight_subgraph.push_back(observer);
+    o.findSubgraph(self, weight_value, weight_subgraph);
```

Comment:
do we need to put this function in ProcessWeightObserverSubgraph?

Reply:
I think it is good to keep all related functions in the class. Serves as an entry point into building and running the graph for each value

Reply:
I feel it's cleaner to make `findSubgraph` a top level function or a member function for InsertQuantDequant, so that `ProcessWeightObserverSubgraph` can just be `SubgraphCloneHelper`